### PR TITLE
app-misc/khard: Compile documentation with sphinx

### PR DESCRIPTION
**Pull request by mjsir911**
Depends on #24385

Having man pages & such is kinda useful.

### CODE REVIEW

**Comment by andrey-utkin** [comment]
Hi Marco!
Thanks for the patches.

One issue is that khard ebuild which you've patched is already marked stable (keywords without the tilde).
To work around that, we just need to leave khard-0.17.0-r1.ebuild intact and create khard-0.17.0-r2.ebuild .
I may rework that myself later.

**Review by andrey-utkin — CHANGES_REQUESTED** [review]
As said, a new revision ebuild must be created.
Also sphinx-autoapi building failed for me, but i will discuss that in PR #24385.

File: app-misc/khard/khard-0.17.0-r1.ebuild [hunk 1]
```diff
@@ -21,6 +21,8 @@ else
 	KEYWORDS="amd64 arm arm64 x86"
 fi
 
+IUSE=doc
```
Comment (thesamesam) [inline app-misc/khard/khard-0.17.0-r1.ebuild]:
```suggestion
IUSE="doc"
```

File: dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	python? ( dev-python/astroid[${PYTHON_USEDEP}] )
+"
+
+DOCS=( README.rst CHANGELOG.rst )
+HTML_DOCS=( html/. )
+
+distutils_enable_tests pytest
+
+src_compile() {
+	sphinx-build -c docs/ -b html docs/ html/
```
Comment (thesamesam) [inline dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild:31]: [You must || die on external commands or you won't know if they failed](https://devmanual.gentoo.org/ebuild-writing/error-handling/#the-die-function):
```suggestion
	sphinx-build -c docs/ -b html docs/ html/ || die
```

File: dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# Copyright 2020 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+PYTHON_COMPAT=( python3_{7..10} )
```
Comment (thesamesam) [inline dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild:6]: Python 3.7 is long-gone. It should give a QA warning on merging.

File: dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+inherit distutils-r1
+
+DESCRIPTION="A new approach to API documentation in Sphinx."
```
Comment (thesamesam) [inline dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild:10]: See proxy maintainers [style guide](https://wiki.gentoo.org/wiki/Project:Proxy_Maintainers/User_Guide/Style_Guide#Descriptive_text):
```suggestion
DESCRIPTION="A new approach to API documentation in Sphinx"
```

File: dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+HOMEPAGE="https://sphinx-autoapi.readthedocs.io/"
+SRC_URI="https://github.com/readthedocs/sphinx-autoapi/archive/v${PV}.tar.gz -> ${P}.tar.gz"
+# SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz" # doesn't include docs/
```
Comment (thesamesam) [inline dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild:13]: You can just put a comment above that "# pypi lacks docs/"

File: app-misc/khard/khard-0.17.0-r2.ebuild [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+		dev-python/sphinx-autoapi[python]
+		dev-python/sphinx-autodoc-typehints
+	)
+"
+
+DOCS=( CHANGES CONTRIBUTING.rst README.md doc/source/examples/khard.conf.example )
+
+distutils_enable_tests setup.py
+
+src_compile() {
```
Comment (thesamesam) [inline app-misc/khard/khard-0.17.0-r2.ebuild:49]: This now shadows distutils-r1_src_compile, you need to call it first.

distutils-r1 exports a src_compile but you're overriding the implementation without calling the one it provides

File: app-misc/khard/khard-0.17.0-r2.ebuild [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+LICENSE="GPL-3"
+SLOT="0"
+
+if [[ "${PV}" == *9999 ]]; then
+	inherit git-r3
+	EGIT_REPO_URI="https://github.com/scheibler/khard"
+else
+	SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+	KEYWORDS="amd64 arm arm64 x86"
```
Comment (thesamesam) [inline app-misc/khard/khard-0.17.0-r2.ebuild:21]: `ekeyword \~all` (see [devmanual](https://devmanual.gentoo.org/keywording/index.html#keywording-on-upgrades))

File: app-misc/khard/khard-0.17.0-r2.ebuild [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+fi
+
+IUSE=doc
```
Comment (thesamesam) [inline app-misc/khard/khard-0.17.0-r2.ebuild:24]:
```suggestion
IUSE="doc"
```

File: app-misc/khard/khard-0.17.0-r2.ebuild [hunk 9]
```diff
@@ -0,0 +1,68 @@
+# Copyright 1999-2021 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+PYTHON_COMPAT=( python3_{8,9} )
+DISTUTILS_USE_SETUPTOOLS=rdepend
```
Comment (thesamesam) [inline app-misc/khard/khard-0.17.0-r2.ebuild:7]: If possible, although it doesn't block this, try:
- EAPI 8
- [PEP517](https://projects.gentoo.org/python/guide/distutils.html#basic-use-pep-517-mode)

These are both nice-to-haves for what you're doing, but the next version bump should try to do it. So, you can treat this as a comment for @andrey-utkin instead if you wish.

File: app-misc/khard/khard-0.17.0-r2.ebuild [hunk 10]
```diff
@@ -0,0 +1,68 @@
+# Copyright 1999-2021 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+PYTHON_COMPAT=( python3_{8,9} )
```
Comment (thesamesam) [inline app-misc/khard/khard-0.17.0-r2.ebuild:6]: Oh, and Python 3.10.

**Comment by mjsir911** [comment]
actually just stumbled onto `distutils_enable_sphinx docs` so gonna use that

**Comment by mjsir911** [comment]
> > actually just stumbled onto `distutils_enable_sphinx docs` so gonna use that
> 
> Oops. I should've thought of that!

Looks like it isn't feasable since other documentation like `man` files are being generated too, but it was a good shot!

File: app-misc/khard/khard-0.17.0-r2.ebuild [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+	dev-python/unidecode[${PYTHON_USEDEP}]
+	dev-python/vobject[${PYTHON_USEDEP}]
+"
+BDEPEND="
+	test? (
+		dev-python/setuptools_scm[${PYTHON_USEDEP}]
+	)
+	doc? (
+		dev-python/sphinx
+		dev-python/sphinx-autoapi[python]
```
Comment (andrey-utkin) [inline app-misc/khard/khard-0.17.0-r2.ebuild:40]: We can avoid optional runtime dep by adding `dev-python/astroid[${PYTHON_USEDEP}]` right there.

**Review by andrey-utkin — CHANGES_REQUESTED** [review]
Thanks for all your work Marco and Sam!

Optional runtime dependency should be avoided in some way.
Keywords dropping from khard should be avoided.

File: dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+# pypi lacks docs/
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="+python" # go javascript dotnet, currently not supported
+
+RDEPEND="
+	dev-python/sphinx[${PYTHON_USEDEP}]
+	python? ( dev-python/astroid[${PYTHON_USEDEP}] )
```
[the remark this answers is not in the document]
Comment (mjsir911) [reply dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild:22]: Hmm, this is a bit tough. From the link, this is basically the case:

>  package being nonfunctional unless at least one of a set of optional runtime dependencies is installed.

`sphinx-autoapi` supports 4 languages: `python`, `go`, `javascript` & `dotnet`. [Currently only python is marked "stable"](https://github.com/readthedocs/sphinx-autoapi#language-support) and that's the language backend I'm interested in with `khard` anyways, so I only added support for that.

I would definitely prefer it as an unconditional dep over not listing it at all and kick the can down the road if the runtime deps *are* actually never needed to be differentiated. 

This optional runtime dependency was sort of just in (half-assed) preparation for that.

File: app-misc/khard/khard-0.17.0-r2.ebuild [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+LICENSE="GPL-3"
+SLOT="0"
+
+if [[ "${PV}" == *9999 ]]; then
+	inherit git-r3
+	EGIT_REPO_URI="https://github.com/scheibler/khard"
+else
+	SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+	KEYWORDS="~amd64 ~x86"
```
[the remark this answers is not in the document]
Comment (thesamesam) [reply app-misc/khard/khard-0.17.0-r2.ebuild:21]: While I generally prefer to drop keywords for new deps so they're tested together, I agree there's no real point here -- so yes, go ahead, but @mjsir911 add <stabilize-allarches/> to metadata.xml?

File: dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild (hunk 12 above)
[the remark this answers is not in the document]
Comment (thesamesam) [reply dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild:22]: It's somewhat sort-of kind-of sometimes okay for Python stuff but I'd probably just list it unconditionally, yes.

**Comment by mjsir911** [comment]
> `FEATURES=test USE="doc" emerge -atv =khard-0.17.0-r2` gives me strange results with slot conflict, but that's not in a clean system so might be irrelevant. Please check that it works if you do any changes.

No slot conflicts from me, some deep dependency loops with `FEATURES=test` but after getting those out of the way mostly everything worked. 

Had to deselect a single test in `dev-python/sphinx-autoapi` for requiring internet access. See readthedocs/sphinx-autoapi#329

**Comment by thesamesam** [comment]
You should do: `emerge -v1 --with-test-deps --onlydeps foo; FEATURES=test emerge -v1 foo` in general to avoid such loops. And throw in USE=doc I suppose after doing that once to verify you don't miss anything :P

File: dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+# pypi lacks docs/
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~arm64 ~x86"
+
+RDEPEND="
+	dev-python/sphinx[${PYTHON_USEDEP}]
+	dev-python/astroid[${PYTHON_USEDEP}]
+"
```
Comment (andrey-utkin) [inline dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild:22]: I would like to be wrong but I'm afraid we need
```
BDEPEND="test? ( $RDEPEND )"
```
Unless I'm missing that this is automatically applied.

"8.1 Dependency Classes" in PMS says src_test phase can count on DEPEND, BDEPEND deps only.
https://projects.gentoo.org/pms/8/pms.html

@thesamesam does that make sense to you?
If so, I may apply this change before merging.

File: app-misc/khard/khard-0.17.0-r2.ebuild [hunk 15]
```diff
@@ -0,0 +1,70 @@
+# Copyright 1999-2021 Gentoo Authors
```
Comment (andrey-utkin) [inline app-misc/khard/khard-0.17.0-r2.ebuild:1]: Year needs updating.
I can update it before merging.

File: dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild [hunk 16]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2020 Gentoo Authors
```
Comment (andrey-utkin) [inline dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild:1]: Year needs updating.
I can update it before merging.

File: dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild (hunk 14 above)
Comment (thesamesam) [reply dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild:22]: d_e_t does this for us

**Comment by andrey-utkin** [comment]
I approve.
I will amend the copyright years and test the change for the last time, and push it.

File: app-misc/khard/khard-0.17.0-r2.ebuild [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+
+	if use doc; then
+		emake -j1 -C doc/ html text man info
+	fi
+}
+
+src_install() {
+	if use doc; then
+		DOCS+=( doc/build/text/ )
+		HTML_DOCS+=( doc/build/html/. )
```
Comment (andrey-utkin) [inline app-misc/khard/khard-0.17.0-r2.ebuild:60]: odd trailing dot, i'll drop that

File: dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+
+RDEPEND="
+	dev-python/sphinx[${PYTHON_USEDEP}]
+	dev-python/astroid[${PYTHON_USEDEP}]
+"
+
+DOCS=( README.rst CHANGELOG.rst )
+
+# Test requires pypi download w/ internet: https://github.com/readthedocs/sphinx-autoapi/issues/329
+EPYTEST_DESELECT="tests/test_integration.py::TestExtensionErrors::test_extension_setup_errors[dotnetexample-override_conf2-AutoAPI"
```
Comment (andrey-utkin) [inline dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild:27]: Seems like the closing bracket is missing.

Comment (thesamesam) [reply dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild:27]: Nit: I'd prefer an array here () instead too

**Review by thesamesam — APPROVED** [review]

**Comment by andrey-utkin** [comment]
By installing sphinx-autoapi with tests enabled, I hit this failure to run tests.
I tried adding dev-python/unidecode to RDEPEND, will see how it goes.
Judging by the backtrace, it's not a test-only dependency.
```
_________________________________________________________________________________________ ERROR collecting tests/python/test_pyintegration.py __________________________________________________________________________________________
ImportError while importing test module '/var/tmp/portage/dev-python/sphinx-autoapi-1.8.4/work/sphinx-autoapi-1.8.4/tests/python/test_pyintegration.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.9/site-packages/_pytest/python.py:578: in _importtestmodule
    mod = import_path(self.fspath, mode=importmode)
/usr/lib/python3.9/site-packages/_pytest/pathlib.py:524: in import_path
    importlib.import_module(module_name)
/usr/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1030: in _gcd_import
    ???
<frozen importlib._bootstrap>:1007: in _find_and_load
    ???
<frozen importlib._bootstrap>:986: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:680: in _load_unlocked
    ???
/usr/lib/python3.9/site-packages/_pytest/assertion/rewrite.py:170: in exec_module
    exec(co, module.__dict__)
tests/python/test_pyintegration.py:14: in <module>
    from autoapi.mappers.python import (
autoapi/__init__.py:3: in <module>
    from .extension import setup
autoapi/extension.py:20: in <module>
    from . import documenters
autoapi/documenters.py:5: in <module>
    from .mappers.python import (
autoapi/mappers/__init__.py:1: in <module>
    from .dotnet import DotNetSphinxMapper
autoapi/mappers/dotnet.py:7: in <module>
    import unidecode
E   ModuleNotFoundError: No module named 'unidecode'
```

**Comment by andrey-utkin** [comment]
I am uneasy with this `autoapi/mappers/dotnet.py`. Isn't our intention to have it disabled?
```
_________________________________________________________________________________________ ERROR collecting tests/python/test_pyintegration.py __________________________________________________________________________________________
ImportError while importing test module '/var/tmp/portage/dev-python/sphinx-autoapi-1.8.4/work/sphinx-autoapi-1.8.4/tests/python/test_pyintegration.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.9/site-packages/_pytest/python.py:578: in _importtestmodule
    mod = import_path(self.fspath, mode=importmode)
/usr/lib/python3.9/site-packages/_pytest/pathlib.py:524: in import_path
    importlib.import_module(module_name)
/usr/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1030: in _gcd_import
    ???
<frozen importlib._bootstrap>:1007: in _find_and_load
    ???
<frozen importlib._bootstrap>:986: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:680: in _load_unlocked
    ???
/usr/lib/python3.9/site-packages/_pytest/assertion/rewrite.py:170: in exec_module
    exec(co, module.__dict__)
tests/python/test_pyintegration.py:14: in <module>
    from autoapi.mappers.python import (
autoapi/__init__.py:3: in <module>
    from .extension import setup
autoapi/extension.py:20: in <module>
    from . import documenters
autoapi/documenters.py:5: in <module>
    from .mappers.python import (
autoapi/mappers/__init__.py:1: in <module>
    from .dotnet import DotNetSphinxMapper
autoapi/mappers/dotnet.py:9: in <module>
    import yaml
E   ModuleNotFoundError: No module named 'yaml'
```

**Comment by andrey-utkin** [comment]
So this made installation of sphinx-autoapi succeed with tests enabled.
```
 RDEPEND="
-       dev-python/sphinx[${PYTHON_USEDEP}]
        dev-python/astroid[${PYTHON_USEDEP}]
+       dev-python/pyyaml[${PYTHON_USEDEP}]
+       dev-python/sphinx[${PYTHON_USEDEP}]
+       dev-python/unidecode[${PYTHON_USEDEP}]
 "
```

File: dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild (hunk 18 above)
Comment (andrey-utkin) [reply dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild:27]: > Seems like the closing bracket is missing.

Turns out there must be no closing bracket - i added it and it made it fail!

**Comment by andrey-utkin** [comment]
Merged! 4584c7cca83960dbe9cdbc7298adc76505a37240 , da660816f0c85116f87b36d18a3a8881a677f80b .
Huge thanks Marco and Sam!

File: app-misc/khard/khard-0.17.0-r2.ebuild (hunk 17 above)
Comment (mjsir911) [reply app-misc/khard/khard-0.17.0-r2.ebuild:60]: btw @andrey-utkin  this is intentional and changes behaviour, without trailing dot HTML documentation is placed in `html/html/` directory, with it it just places it in `html/`

File: dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild (hunk 18 above)
Comment (mjsir911) [reply dev-python/sphinx-autoapi/sphinx-autoapi-1.8.4.ebuild:27]: The full test's pytest parameter is something stupid that I didn't want to drop the whole thing in:
```
tests/test_integration.py::TestExtensionErrors::test_extension_setup_errors[dotnetexample-override_conf2-AutoAPI of type `dotnet` requires following packages to be installed and included in extensions list: sphinxcontrib.dotnetdomain (available as "sphinxcontrib-dotnetdomain" on PyPI)]
```

File: app-misc/khard/khard-0.17.0-r2.ebuild (hunk 17 above)
Comment (mjsir911) [reply app-misc/khard/khard-0.17.0-r2.ebuild:60]: Although I suppose you would want a trailing dot after `text/` as well if you don't want documentation showing up in `doc/text/`

**Comment by mjsir911** [comment]
Thank you Andrey for finalizing this and tracking down the additional deps!

File: app-misc/khard/khard-0.17.0-r2.ebuild (hunk 17 above)
Comment (andrey-utkin) [reply app-misc/khard/khard-0.17.0-r2.ebuild:60]: Thanks for the clarification.
I've pushed such changes in 174a41f63e7

**Pull request closed without merging**